MantidPlot/src/ProjectRecoveryGUIs/ProjectRecoveryModel.cpp

```diff
@@ -0,0 +1,219 @@
+  m_recoveryRunning = true;
+  m_presenter->changeStartMantidToCancelLabel();
+  // Clear the ADS
+  Mantid::API::AnalysisDataService::Instance().clear();
+
+  // Recovery given the checkpoint selected here
+  auto stringSpacePos = selected.find(" ");
+  if (stringSpacePos != std::string::npos)
+    selected.replace(stringSpacePos, 1, "T");
+  Poco::Path checkpoint(m_projRec->getRecoveryFolderLoadPR());
+  checkpoint.append(selected);
+  Poco::Path output(Mantid::Kernel::ConfigService::Instance().getAppDataDir());
+  output.append("ordered_recovery.py");
+
+  m_projRec->openInEditor(checkpoint, output);
+  createThreadAndManage(checkpoint);
+
+  selected.replace(selected.find("T"), 1, " ");
+  if (m_failedRun) {
+    updateCheckpointTried(selected);
+  }
+  m_recoveryRunning = false;
+  // Close View
+  m_presenter->closeView();
+}
+void ProjectRecoveryModel::openSelectedInEditor(std::string &selected) {
+  m_recoveryRunning = true;
+  // Clear the ADS
+  Mantid::API::AnalysisDataService::Instance().clear();
+
+  // Open editor for this checkpoint
+  auto stringSpacePos = selected.find(" ");
+  if (stringSpacePos != std::string::npos)
+    selected.replace(stringSpacePos, 1, "T");
+  auto beforeCheckpoint = m_projRec->getRecoveryFolderLoadPR();
+  Poco::Path checkpoint(beforeCheckpoint);
+  checkpoint.append(selected);
+  Poco::Path output(Mantid::Kernel::ConfigService::Instance().getAppDataDir());
+  output.append("ordered_recovery.py");
+
+  m_projRec->openInEditor(checkpoint, output);
+  // Restart project recovery as we stay synchronous
+  m_projRec->clearAllCheckpoints(beforeCheckpoint);
+  m_projRec->startProjectSaving();
+
+  selected.replace(selected.find("T"), 1, " ");
+  if (m_failedRun) {
+    updateCheckpointTried(selected);
+  }
+  m_recoveryRunning = false;
+  m_failedRun = false;
+  // Close View
+  m_presenter->closeView();
+}
+
+void ProjectRecoveryModel::fillRow(Poco::Path path,
+                                   std::string checkpointName) {
+  std::string lengthOfFile = findNumberOfWorkspacesInDirectory(path);
+  std::string checked = "No";
+  std::vector<std::string> nextVector = {checkpointName, lengthOfFile, checked};
+  m_rows.emplace_back(nextVector);
+}
+
+void ProjectRecoveryModel::fillFirstRow() {
+  auto paths = m_projRec->getListOfFoldersInDirectoryPR(
+      m_projRec->getRecoveryFolderLoadPR());
+
+  std::sort(paths.begin(), paths.end(), [](auto &a, auto &b) {
+    Poco::File a1(a);
+    Poco::File b1(b); // Last modified is first!
+    return a1.getLastModified() > b1.getLastModified();
+  });
+
+  // Grab the last path as that is the one that should be loaded
+  auto path = paths[0];
+  std::string checkpointName = path.directory(path.depth() - 1);
+  checkpointName.replace(checkpointName.find("T"), 1, " ");
+  fillRow(path, checkpointName);
+}
+
+void ProjectRecoveryModel::fillRows() {
+  auto paths = m_projRec->getListOfFoldersInDirectoryPR(
```

Comment:
`const`

Reply:
Paths changes later in the code, so can't const.